crates/unftp-auth-json/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+//! [`Authenticator`] implementation that authenticates against a JSON file.
+//!
+//! [`Authenticator`]: libunftp::auth::Authenticator
+
+use async_trait::async_trait;
+use libunftp::auth::{AuthenticationError, Authenticator, DefaultUser};
+use ring::{
+    digest::SHA512_OUTPUT_LEN,
+    pbkdf2::{verify, PBKDF2_HMAC_SHA512},
+};
+use serde::Deserialize;
+use std::{
+    collections::{BTreeSet, HashMap},
+    convert::TryInto,
+    fs,
+    num::NonZeroU32,
+    path::Path,
+    time::Duration,
+};
+use tokio::time::sleep;
+
+#[derive(Deserialize, Clone, Debug)]
+struct Credential {
+    username: String,
+    pbkdf2_salt: String,
+    pbkdf2_key: String,
+    pbkdf2_iter: NonZeroU32,
+}
+
+/// [`Authenticator`](libunftp::auth::Authenticator) implementation that authenticates against a JSON.
+///
+/// Verifies passwords against pbkdf2_key using the corresponding parameters form JSON.
+/// Example credentials file format:
+/// [
+//   {
+//     "username": "testuser1",
+//     "pbkdf2_salt": "testuser1.acme.bol.com",
+//     "pbkdf2_key": "<<BASE_64_KDF>>",
+//     "pbkdf2_iter": 500000
+//   },
+//   {
+//     "username": "testuser2",
+//     "pbkdf2_salt": "testuser2.acme.bol.com",
+//     "pbkdf2_key": "<<BASE_64_KDF>>",
+//     "pbkdf2_iter": 500000
+//   }
+// ]
+
+#[derive(Clone, Debug)]
+pub struct JsonAuthenticator {
+    db: HashMap<String, Password>,
+}
+
+#[derive(Clone, Debug)]
+struct Password {
+    pbkdf2_salt: String,
+    pbkdf2_key: [u8; SHA512_OUTPUT_LEN],
```

Comment:
It would be better to support variable length keys. It is common for pbkdf2 password utilities to support a length (`-l <n>` for nettle-pbkdf2 and `-keylen` for openssl 3.0).

Reply:
Ok, nvm, the PBKDF2 standard says not to use length larger than digest output length